onnxruntime/test/providers/qnn/gemm_op_test.cc

```diff
@@ -0,0 +1,341 @@
+
+    for (const auto& attr : attrs) {
+      gemm_node.AddAttributeProto(attr);
+    }
+
+    // op_output -> Q -> DQ -> output
+    AddQDQNodePairWithOutputAsGraphOutput<InputAQType>(builder, gemm_output, output_qparams[0].scale,
+                                                       output_qparams[0].zero_point, use_contrib_qdq);
+  };
+}
+
+// Runs a QDQ Gemm model on the QNN (HTP) EP and the ORT CPU EP. Checks the graph node assignment and that inference
+// running the QDQ model on QNN EP is at least as accurate as on ORT CPU EP (compared to the baseline float32 model).
+template <typename InputAQType, typename InputBQType>
+static void RunQDQGemmTestOnHTP(const std::vector<TestInputDef<float>>& input_defs,
+                                const std::vector<ONNX_NAMESPACE::AttributeProto>& attrs,
+                                ExpectedEPNodeAssignment expected_ep_assignment,
+                                int opset = 13,
+                                float f32_abs_err = 1e-4f,
+                                bool use_contrib_qdq = false) {
+  ProviderOptions provider_options;
+
+#if defined(_WIN32)
+  provider_options["backend_path"] = "QnnHtp.dll";
+#else
+  provider_options["backend_path"] = "libQnnHtp.so";
+#endif
+  auto f32_model_builder = BuildOpTestCase<float>("Gemm", input_defs, {}, attrs);
+  auto qdq_model_builder = BuildQDQGemmTestCase<InputAQType, InputBQType>(input_defs, attrs, use_contrib_qdq);
+  TestQDQModelAccuracy<InputAQType>(f32_model_builder,
+                                    qdq_model_builder,
+                                    provider_options,
+                                    opset,
+                                    expected_ep_assignment,
+                                    f32_abs_err);
+}
+
+// Test 8-bit QDQ Gemm with dynamic inputs A and Bias. The B input is an initializer.
+TEST_F(QnnHTPBackendTests, Gemm_Dynamic_A_Static_B_Dynamic_Bias_U8) {
+  std::vector<float> input_a_data = GetFloatDataInRange(-10.0f, 10.0f, 6);
+  std::vector<float> input_b_data = GetFloatDataInRange(-5.0f, 5.0f, 24);
+  std::vector<float> input_c_data = GetFloatDataInRange(-1.0f, 1.0f, 4);
+  RunQDQGemmTestOnHTP<uint8_t, uint8_t>({TestInputDef<float>({1, 6}, false, input_a_data),
+                                         TestInputDef<float>({6, 4}, true, input_b_data),
+                                         TestInputDef<float>({1, 4}, false, input_c_data)},
+                                        {},
+                                        ExpectedEPNodeAssignment::All);
+}
+
+// Test 16-bit QDQ Gemm with dynamic inputs A and Bias. The B input is an initializer.
+// TODO: Inaccuracy detected for output 'output_0', element 0.
+// Output quant params: scale=0.001872879103757441, zero_point=0.
+// Expected val: 120.73912048339844
+// QNN QDQ val: 0 (err 120.73912048339844)
+// CPU QDQ val: 120.73889923095703 (err 0.00022125244140625)
+TEST_F(QnnHTPBackendTests, DISABLED_Gemm_Dynamic_A_Static_B_Dynamic_Bias_U16) {
```

Comment:
>DISABLED_Gemm_Dynamic_A_Static_B_Dynamic_Bias_U16 [](start = 27, length = 49)

maybe also something we can improve in our GemmOpBuilder to swap A & B.

Reply:
Let's check what QNN can do first.

---
In reply to: [1331935609](https://github.com/microsoft/onnxruntime/pull/17457#discussion_r1331935609) [](ancestors = 1331935609)